src/btree/bt_page.c

```diff
@@ -635,12 +645,61 @@ __inmem_col_int(WT_SESSION_IMPL *session, WT_PAGE *page)
         ref = *refp++;
         ref->home = page;
         ref->pindex_hint = hint++;
-        ref->addr = unpack.cell;
-        ref->ref_recno = unpack.v;
+
+        if (first && unpack.v != page_recno) {
+            /*
+             * There's a gap in the namespace. Create a deleted leaf page (with no address) to cover
+             * that gap. We allocated an extra slot in the array above to make room for this case.
+             * (Note that this doesn't result in all gaps being covered, just ones on the left side
+             * of the tree where we need to be able to search to them. Other gaps end up covered by
+             * the insert list of the preceding leaf page.)
+             */
+            ref->addr = NULL;
+            ref->ref_recno = page_recno;
+            F_SET(ref, WT_REF_FLAG_LEAF);
+            WT_REF_SET_STATE(ref, WT_REF_DELETED);
+            gap = true;
+
+            /* Get the next ref. */
+            ref = *refp++;
+            ref->home = page;
+            ref->pindex_hint = hint++;
+        }
+        first = false;
 
         F_SET(ref, unpack.type == WT_CELL_ADDR_INT ? WT_REF_FLAG_INTERNAL : WT_REF_FLAG_LEAF);
+
+        if (unpack.type == WT_CELL_ADDR_DEL) {
+            /*
+             * If a page was deleted without being read (fast truncate), and the delete committed,
+             * but older transactions in the system required the previous version of the page to
+             * remain available or the delete can still be rolled back by RTS, a deleted-address
+             * type cell is written. We'll see that cell on a page if we read from a checkpoint
+             * including a deleted cell or if we crash/recover and start off from such a checkpoint.
+             * Recreate the fast-delete state for the page.
+             */
+            if (F_ISSET(page->dsk, WT_PAGE_FT_UPDATE)) {
+                WT_RET(__wt_calloc_one(session, &ref->ft_info.del));
+                *ref->ft_info.del = unpack.page_del;
+            }
+            WT_REF_SET_STATE(ref, WT_REF_DELETED);
+        }
+        ref->addr = unpack.cell;
+        ref->ref_recno = unpack.v;
     }
     WT_CELL_FOREACH_END;
+
+    if (gap == false) {
```

Comment:
I'm not sure how common it is, so I was hesitant to do that. Reallocating the array and allocating a new ref is probably no worse than allocating an extra ref up front and freeing it, but copying the array in the reallocation isn't free. Probably it doesn't matter.

The root problem is that wt_page_alloc happens before we can tell if we're going to need the extra ref, but changing that around didn't seem like a good idea either.

(IOW, I didn't like any of the options and chose the one I thought probably least bad...)

Let me chew on it a bit.

Reply:
I went to try it and discovered a probably-fatal complication: the ref array is part of the same memory block as the page, along with the WT_PAGE_INDEX, so reallocating it will make kind of a mess.

I guess another approach is to inspect the first cell at line 374 and ask for another entry conditionally there. That also seemed ugly, but it's relatively self-contained...

Reply:
Thanks for trying. I'd appreciate if you can try checking the cell, but if that appears worse, I'll be content with it as-is.

Reply:
It seems better. Coming shortly.

Reply:
Pushed in [4d5ad9f](https://github.com/wiredtiger/wiredtiger/pull/8129/commits/4d5ad9ff5cab424d4c9d2755bd925a497e356ab7).